# Урок 5

# Сингулярное разложение матриц

### __1.__ Найти с помощью NumPy SVD для матрицы

### $$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

### __Решение:__

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [2]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [3]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


In [4]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]


In [5]:
# Убедимся, что она действительно ортогональна
print(f'{np.dot(U.T, U)}')

[[ 1. -0.  0.  0.  0.]
 [-0.  1. -0. -0. -0.]
 [ 0. -0.  1.  0. -0.]
 [ 0. -0.  0.  1. -0.]
 [ 0. -0. -0. -0.  1.]]


In [6]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [7]:
# Убедимся, что она действительно ортогональна
print(f'{np.dot(V.T, V)}')

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [8]:
# Проведем проверку
print(f'{np.dot(np.dot(U, D), V.T)}\n=\n{A}')

[[ 1.  2. -0.]
 [-0.  0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [ 0.  1.  0.]]
=
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


### __Ответ:__ $A=\begin{pmatrix}
0.17 & 0.16 & -0.53 & -0.8 & -0.16\\
0.39 & -0.53 & 0.61 & -0.43 & 0.03\\
-0.14 & -0.82 & -0.52 & 0.14 & 0.07\\
0.89 & 0.06 & -0.25 & 0.38 & -0.06\\
0.08 & 0.11 & -0.08 & -0.11 & 0.98
\end{pmatrix}\begin{pmatrix}
8.82 & 0 & 0\\
0 & 6.14 & 0\\  
0 & 0 & 2.53\\
0 & 0 & 0\\
0 & 0 & 0
\end{pmatrix}\begin{pmatrix}
0.07 & -0.37 & -0.93\\
0.72 & 0.67 & -0.21\\
0.69 & -0.65 & 0.31
\end{pmatrix}^T$

### __2.__ Для матрицы из предыдущего задания найти:

### а) евклидову норму;
    
### б) норму Фробениуса.

### __Решение:__

### а)

Евклидова норма матрицы равна

### $$\left \| A \right \|_{E}=\mu_{1}.$$

In [9]:
print(f'Евклидова норма по максимальному сингулярному числу:\n{s[0]}')

print(f'Евклидова норма методом np.linalg.norm:\n{np.linalg.norm(A, ord = 2)}')

Евклидова норма по максимальному сингулярному числу:
8.824868854820448
Евклидова норма методом np.linalg.norm:
8.824868854820442


### б)

Норма Фробениуса вычисляется как

### $$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

In [10]:
print(f'Норма Фробениуса по формуле:\n{np.sqrt(np.sum(s**2))}')

print(f'Норма Фробениуса методом np.linalg.norm:\n{np.linalg.norm(A)}')

Норма Фробениуса по формуле:
11.045361017187265
Норма Фробениуса методом np.linalg.norm:
11.045361017187261


### __Ответ:__ а); 8.824868854820448 б) 11.045361017187265.